### DISCUSSION TITLE
Support for dotnet-script?

### DISCUSSION THREAD

**Discussion by lispercat** [Olical/conjure#604] [General]
Thanks for the great plugin! I wonder how much effort would be to add support for dotnet-script? 
It's a great repl environment for c#

**Comment by Olical** [comment]
I have no experience with this but I'd be happy to help get things going! In the middle of a large migration right now that we're wrapping up, lots of stuff to do but I'll keep this on my radar.

I'm hoping to make it a lot easier to throw together clients of various kinds now that it's easier to work with the source code of Conjure. I'm hoping more people will be able to jump in and self service since all you need to do is add https://github.com/Olical/nfnl now and you can just edit Conjure's `.fnl` source files without thinking about compiling them.

There's also no magic syntax or macros now, it's just plain Fennel which should be a lot easier for people to pick up.

So, this is to say: I don't see why not! But I probably won't jump on this right away. I'll do what I can to help make this a reality 😃 

Until then, please check out the new `main` branch which contains the large conversion work, this version will eventually replace the `master` branch but that will be kept around for those that can't or won't switch away (despite there being no difference for the user).

**Comment by russtoku** [comment]
I'm not at all familiar with C# and dotnet-script so I'm guessing that this might work.

If you'd like to try a quick proof-of-concept, once you have the Conjure plug-in configured for your Neovim, you can try using the `snd-s7` client to see how running a `dotnet-script` REPL and sending it C# code from a buffer.

Here's how:
- Using the `snd-s7` client means that you'll edit a file with an extension of `.scm` instead of `.csx`. This is just a temporary set-up and a way to use the `snd-s7` client.
- After starting Neovim and before creating or loading a file for testing, set up Conjure to use the `snd-s7` client for files with an extension of `.scm` and to use `dotnet-script` as the command to run the REPL. Run these on the Neovim command line:
```
:lua vim.g["conjure#filetype#scheme"] = "conjure.client.snd-s7.stdio"
:lua vim.g["conjure#client#snd-s7#stdio#command"] = "dotnet-script"
```
- Create a file to test with. Let's say something like `test.scm`. From the Neovim command line, you can use `:e test.scm`.
  - This should have Conjure start a REPL using the `dotnet-script` command.
  - You'll have an empty file loaded if this is the first time and you didn't save things before.
- Add some C# code to the buffer.
- Send a C# statement to the REPL by selecting what you want and then typing`,E`. You should see the results of the evaluation in the HUD (heads-up display) pop-up window and in-lined after the selected code.

If that works, you can try using the `snd-s7` client code in `fnl/conjure/client/snd-s7/stdio.fnl` as a start for a `fnl/conjure/client/dotnet/stdio.fnl` or `fnl/conjure/client/dotnet-script/stdio.fnl`.

Hope this helps!

**Reply by russtoku** [reply]
I'll have to take back what I said.

After trying it out with the [`tio`](https://github.com/tio/tio) command that connects to microcontroller devices via the USB serial port, I can definitely say that it won't work.

You can try using the `sql` or `snd-s7` clients as a starting point to write a C++ client with `dotnet-script` as the REPL command. You might want to implement it as `fnl/conjure/client/cpp/stdio.fnl`. If you do use either of `sql` or `snd-s7` clients as a starting point, please note that they use `fnl/conjure/remote/stdio-rt.fnl` and not `fnl/conjure/remote/stdio.fnl`. The difference has to do with how requests and responses with the REPL are handled.
